File: Week 3/Week 3 Template/Examples/data_structures_example.cpp
#include <bits/stdc++.h>
using namespace std;

void queue_example(){
    queue<int> q; // Nothing else needed to initialize a queue

    // We can push elements to the back of the queue with push()
    q.push(3);
    q.push(1);
    q.push(4);

    // We can pop elements from the front of the queue with pop()
    q.pop();

    // We can get the front element with front()
    cout << q.front() << endl; // 1
    q.pop();
    cout << q.front() << endl; // 4
    q.pop();

    // We can check if the queue is empty with empty()
    if(q.empty()) cout << "The queue is empty" << endl;
    else cout << "The queue is not empty" << endl;

    // We can get the size of the queue with size()
    q.push(2);
    q.push(7);
    q.push(1);
    q.push(8);
    cout << "The size of the queue is " << q.size() << endl;
}

void stack_example(){
    // Actually we can just use a vector, which has the added bonus of random access
    // But C++ does have a stack class:
    stack<int> s; 

    // We can push elements to the top of the stack with push()
    s.push(3);
    s.push(1);
    s.push(4);

    // We can pop elements from the top of the stack with pop()
    s.pop();

    // We can get the top element with top()
    cout << s.top() << endl; // 1

    // We can check if the stack is empty with empty()
    if(s.empty()) cout << "The stack is empty" << endl;
    else cout << "The stack is not empty" << endl;

    // We can get the size of the stack with size()
    s.push(4);
    s.push(5);
    s.push(6);
    cout << "The size of the stack is " << s.size() << endl;
}

void deque_example(){
    // A deque is a double ended queue, which means we can push and pop from both ends
    deque<int> d;

    // We can push elements to the back of the deque with push_back()
    d.push_back(3);
    d.push_back(1);
    d.push_back(4);

    // We can pop elements from the back of the deque with pop_back()
    d.pop_back();

    // We can push elements to the front of the deque with push_front()
    d.push_front(2);
    d.push_front(7);
    d.push_front(1);

    // We can pop elements from the front of the deque with pop_front()
    d.pop_front();

    // We can get the front element with front()
    cout << d.front() << endl; // 7

    // We can get the back element with back()
    cout << d.back() << endl; // 1

    // We can check if the deque is empty with empty()
    if(d.empty()) cout << "The deque is empty" << endl;
    else cout << "The deque is not empty" << endl;

    // We can get the size of the deque with size()
    cout << "The size of the deque is " << d.size() << endl;

    // deques support random access, so we can get the ith element with []
    for(int i = 0; i < d.size(); i++){
        cout << d[i] << " ";
    }
    cout << endl;
}

void linked_list_example(){
    // If the normal queue, stack, and deque are not enough, we can use a linked list
    // C++ has a built in linked list class called list, but if we want something more general
    // We can implement our own linked list using indices in a vector
    int n = 10;
    vector<int> next(n + 1);
    vector<int> prev(n + 1);
    // For a doubly linked list, we can store the next and previous nodes for each index
    for(int i = 1; i <= n; i++){
        if(i != n) next[i] = i + 1;
        if(i != 1) prev[i] = i - 1;
    }
    
    // (If you haven't seen this, this is a lambda function)
    // (It lets us define a function inside another function)
    auto remove = [&](int i){
        // Remove the ith element from the linked list
        // This can be done by making the neighbors of the ith element point to each other
        next[prev[i]] = next[i];
        prev[next[i]] = prev[i];
    };
    auto insert_after = [&](int i, int j){
        // Insert the jth element after the ith element
        // This can be done by making j point to the element after i, and then fixing all the other pointers
        next[j] = next[i];
        prev[j] = i;
        next[i] = j;
        prev[next[j]] = j;
    };

    remove(5);
    remove(8);
    remove(9);

    insert_after(2,9);
    
    // We can iterate over the whole list by following the next pointers
    int start = 1;
    while(start != 0){
        cout << start << " ";
        start = next[start];
    }
    cout << endl;

    // For more complex problems, we can keep track of different properties of the nodes
    // And have them point to nodes with other relationships
}

int main(){
    
    queue_example();
    stack_example();
    deque_example();
    linked_list_example();  
    
}

File: Week 3/Week 3 Template/Examples/data_structures_example.java
import java.io.*;
import java.util.*;

class Main{
    static void queueExample(){
        Queue<Integer> q = new LinkedList<>(); // We can declare a queue using the LinkedList class
        q.add(3); // Add an element to the end of the queue
        q.add(1);
        q.add(4);

        System.out.println(q.peek()); // Peek at the front of the queue

        System.out.println(q.poll()); // Remove and return the front of the queue

        System.out.println("The size of the queue is " + q.size()); // Get the size of the queue

        if(q.isEmpty()){  // Check if the queue is empty
            System.out.println("The queue is empty");
        }
        else{
            System.out.println("The queue is not empty");

        }
    }
    static void stackExample(){
        // We can use an ArrayList as a stack, but Java has a Stack class, so I'll show that here
        Stack<Integer> s = new Stack<>(); // We can declare a stack using the Stack class
        s.push(3); // Add an element to the top of the stack
        s.push(1);
        s.push(4);

        System.out.println(s.peek()); // Peek at the top of the stack

        System.out.println(s.pop()); // Remove and return the top of the stack

        System.out.println("The size of the stack is " + s.size()); // Get the size of the stack

        if(s.isEmpty()){  // Check if the stack is empty
            System.out.println("The stack is empty");
        }
        else{
            System.out.println("The stack is not empty");

        }
    }
    static void dequeExample(){
        Deque<Integer> d = new ArrayDeque<>(); // We can declare a deque using the ArrayDeque class
        d.addLast(3); // Add an element to the end of the deque
        d.addLast(1);
        d.addLast(4);

        System.out.println(d.peekLast()); // Peek at the end of the deque

        d.addFirst(2); // Add an element to the front of the deque
        d.addFirst(7);
        d.addFirst(1); 

        System.out.println(d.peekFirst()); // Peek at the front of the deque

        System.out.println(d.pollLast()); // Remove and return the end of the deque

        System.out.println(d.pollFirst()); // Remove and return the front of the deque

        System.out.println("The size of the deque is " + d.size()); // Get the size of the deque

        if(d.isEmpty()){  // Check if the deque is empty
            System.out.println("The deque is empty");
        }
        else{
            System.out.println("The deque is not empty");

        }

        // ArrayDeque doesn't directly support random access
        // But we can call toArray() to get an array representation of the deque
        // This takes O(n) per call; if you need to this a lot, you should probably implement your own deque
        Integer[] arr = d.toArray(new Integer[0]);
        System.out.println(Arrays.toString(arr));
    }
    static void linkedListExample(){
        // If the normal queue, stack, and deque are not enough, we can use a linked list
        // Java has a built in linked list class called list, but if we want something more general
        // We can implement our own linked list using indices in an array
        int n = 10;
        
        // For a doubly linked list, we can store the next and previous nodes for each index

        int[] next = new int[n + 1];
        int[] prev = new int[n + 1];
        
        for(int i = 1; i <= n; i++){
            if(i != n) next[i] = i + 1;
            if(i != 1) prev[i] = i - 1;
        }

        remove(5, next, prev); 
        remove(8, next, prev);
        remove(9, next, prev);

        insertAfter(2, 9, next, prev);


        // We can iterate over the whole list by following the next pointers
        int start = 1;
        while(start != 0){
            System.out.print(start + " ");
            start = next[start];
        }
        System.out.println();

        
        // For more complex problems, we can keep track of different properties of the nodes
        // And have them point to nodes with other relationships
    }
    static void remove(int x, int[] next, int[] prev){
        // This let's us remove x from the array list, by having the neighbors of x point to each other
        next[prev[x]] = next[x];
        prev[next[x]] = prev[x];
    }
    static void insertAfter(int x, int y, int[] next, int[] prev){
        // We can insert y after x by having y point to the node after x, and then fixing all the other pointers
        next[y] = next[x];
        prev[y] = x;
        next[x] = y;
        prev[next[y]] = y;
    }
    public static void main(String[] args) throws Exception{
    	queueExample();
    	stackExample();
    	dequeExample();
    	linkedListExample();
    }	
}
